Compiler bug: C error: '{' expected (got ";")

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->
**V doctor:**
```
V full version: V 0.4.3 2964855.c7ce72f
OS: linux, Debian GNU/Linux 12 (bookworm)
Processor: 8 cpus, 64bit, little endian, Intel(R) Core(TM) i5-8265U CPU @ 1.60GHz

getwd: /home/leo/Documents/floats
vexe: /opt/v/v
vexe mtime: 2023-12-23 16:09:23

vroot: OK, value: /opt/v
VMODULES: OK, value: /home/leo/.vmodules
VTMP: OK, value: /tmp/v_1000

Git version: git version 2.39.2
Git vroot status: weekly.2023.51-28-gc7ce72f3
.git/config present: true

CC version: cc (Debian 12.2.0-14) 12.2.0
thirdparty/tcc status: thirdparty-linux-amd64 12f392c3-dirty

```

**What did you do?**
`v -g -o vdbg cmd/v && vdbg src/write.v`
```v
import runtime { is_big_endian }

union F32A {
	float f32
	value [4]u8
}

union F64A {
	float f64
	value [8]u8
}

pub fn read_f32le(bytes [4]u8) f32 {
	if is_big_endian() {
		bytes_rev := [4]u8{init: bytes[3 - index]}
		float := F32A{value: bytes_rev}
		unsafe {
			return float.float
		}
	} else {
		float := F32A{value: bytes}
		unsafe {
			return float.float
		}
	}
}

pub fn write_f32le(float f32) [4]u8 {
	float_copy := F32A{float: float}
	if is_big_endian() {
		unsafe {
			return [4]u8{init: float_copy.value[3 - index]}
		}
	} else {
		unsafe {
			return float_copy.value
		}
	}
}
```

**What did you expect to see?**

nothing

**What did you see instead?**
```
==================
/tmp/v_1000/write.10246023703988586250.tmp.c:18219: error: '{' expected (got ";")
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

```
> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

I get another error on Windows 11:
```
==================
C:/Users/mclr/AppData/Local/Temp/v_0/test.5950236487262907293.tmp.c:17619: warning: implicit declaration of function 'CreateSymbolicLinkW'
C:/Users/mclr/AppData/Local/Temp/v_0/test.5950236487262907293.tmp.c:18596: error: character array initializer must be a literal, optionally enclosed in braces
...
==================
(Use `v -cg` to print the entire error message)

builder error:
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang
```

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -5282,11 +5282,17 @@ fn (mut g Gen) return_stmt(node ast.Return) {
 					if node.exprs[0] is ast.Ident {
 						g.write('memcpy(${tmpvar}.ret_arr, ${g.expr_string(node.exprs[0])}, sizeof(${g.typ(node.types[0])})) /*ret*/')
 					} else if node.exprs[0] in [ast.ArrayInit, ast.StructInit] {
-						tmpvar2 := g.new_tmp_var()
-						g.write('${g.typ(node.types[0])} ${tmpvar2} = ')
-						g.expr_with_cast(node.exprs[0], node.types[0], g.fn_decl.return_type)
-						g.writeln(';')
-						g.write('memcpy(${tmpvar}.ret_arr, ${tmpvar2}, sizeof(${g.typ(node.types[0])})) /*ret*/')
+						if node.exprs[0] is ast.ArrayInit && node.exprs[0].is_fixed {
+							g.write('memcpy(${tmpvar}.ret_arr, ')
+							g.expr_with_cast(node.exprs[0], node.types[0], g.fn_decl.return_type)
+							g.write(', sizeof(${g.typ(node.types[0])})) /*ret*/')
+						} else {
+							tmpvar2 := g.new_tmp_var()
+							g.write('${g.typ(node.types[0])} ${tmpvar2} = ')
+							g.expr_with_cast(node.exprs[0], node.types[0], g.fn_decl.return_type)
+							g.writeln(';')
+							g.write('memcpy(${tmpvar}.ret_arr, ${tmpvar2}, sizeof(${g.typ(node.types[0])})) /*ret*/')
+						}
 					} else {
 						g.write('memcpy(${tmpvar}.ret_arr, ')
 						g.expr_with_cast(node.exprs[0], node.types[0], g.fn_decl.return_type)

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -5282,7 +5282,8 @@ fn (mut g Gen) return_stmt(node ast.Return) {
 					if node.exprs[0] is ast.Ident {
 						g.write('memcpy(${tmpvar}.ret_arr, ${g.expr_string(node.exprs[0])}, sizeof(${g.typ(node.types[0])})) /*ret*/')
 					} else if node.exprs[0] in [ast.ArrayInit, ast.StructInit] {
-						if node.exprs[0] is ast.ArrayInit && node.exprs[0].is_fixed {
+						if node.exprs[0] is ast.ArrayInit && node.exprs[0].is_fixed
+							&& node.exprs[0].has_init {
 							g.write('memcpy(${tmpvar}.ret_arr, ')
 							g.expr_with_cast(node.exprs[0], node.types[0], g.fn_decl.return_type)
 							g.write(', sizeof(${g.typ(node.types[0])})) /*ret*/')

--- a/vlib/v/tests/fixed_array_return_decl_test.v
+++ b/vlib/v/tests/fixed_array_return_decl_test.v
@@ -1,0 +1,58 @@
+import runtime { is_big_endian }
+
+union F32A {
+	float f32
+	value [4]u8
+}
+
+union F64A {
+	float f64
+	value [8]u8
+}
+
+pub fn read_f32le(bytes [4]u8) f32 {
+	if is_big_endian() {
+		bytes_rev := [4]u8{init: bytes[3 - index]}
+		float := F32A{
+			value: bytes_rev
+		}
+		unsafe {
+			return float.float
+		}
+	} else {
+		float := F32A{
+			value: bytes
+		}
+		unsafe {
+			return float.float
+		}
+	}
+}
+
+pub fn write_f32le(float f32) [4]u8 {
+	float_copy := F32A{
+		float: float
+	}
+	if is_big_endian() {
+		unsafe {
+			return [4]u8{init: float_copy.value[3 - index]}
+		}
+	} else {
+		unsafe {
+			return float_copy.value
+		}
+	}
+}
+
+fn test_zeroed() {
+	a := [4]u8{}
+	b := write_f32le(0.0)
+	assert a == b
+}
+
+fn test_value() {
+	mut a2 := [4]u8{}
+	a2 = [u8(205), 204, 204, 61]!
+	b2 := write_f32le(0.1)
+	assert a2 == b2
+}

